### ISSUE TITLE
[$250] Chat - Scrolling to top of page sending a message scrolls down to previous message

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#53685]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** V. 9.0.72-0
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** Yes, reproducible on both
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N
**Email or phone of affected tester (no customers):** N
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Launch app in both mweb and app
2. Open a chat with long chat history
3. Scroll to the top of the page
4. Send a message

## Expected Result:
Scrolling to top of page sending a message must scroll down to latest message .

## Actual Result:
In mweb, scrolling to top of page sending a message scrolls down to latest message but in Android previously sent message is briefly shown before latest message shown

## Workaround:
Unknown

## Platforms:
- [x] Android: Standalone
- [x] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos
https://github.com/user-attachments/assets/c2f647d2-ed9c-40a0-88fa-14470665e01e

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021866170377018809152</li>
        <li>Upwork Job ID: 1866170377018809152</li>
        <li>Last Price Increase: 2024-12-09</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @rinej </details>

**Comment by rinej** [comment]
Hello, I am Tomek from Callstack, I would like to work on that issue

**Comment by rinej** [comment]
I am working on the Draft PR with proposal, I will share it shortly

**Comment by rinej** [comment]
**Problem** 
When we receive a new message, we instantly scroll to the bottom. In some cases, especially on Android or slow-performing devices, sending a message takes longer than scrolling to the bottom. This creates the impression that we are scrolling to a previous message.

**Solution**
In the PR, we add behavior to ensure that the chat only scrolls to the bottom if the newly sent message is actually visible to the user.

As stated in the draft PR, it will also be worth investigating the performance of sending messages, because on some mid-range Android phones with high-traffic users, there is visible lag when sending a message.

Here is the PR -> https://github.com/Expensify/App/pull/54071

**Comment by rinej** [comment]
Thanks for the input! I tested this in offline mode. The pusher event is triggered when we send a message from the same account we're logged into; however, the message won't be sent and will remain offline. That said, the user experience is quite similar to when we are online. Here are some example videos:

https://github.com/user-attachments/assets/6f09c0ed-f9e4-427e-9d90-8b58288d984a


https://github.com/user-attachments/assets/9fbaa776-e4ec-4b1e-9ce9-7a536a5cdedb



For the RCA, the primary concern is that the event triggered when sending a message occurs immediately. However, there is complex logic and computation involved before displaying the new message. In particular, the getOrderedReportID method takes some time to complete, especially on mid-range Android devices. This results in a discrepancy between the moment the message is sent and when it appears on the screen.

To address this, in this ticket, we propose scrolling only when the message becomes visible, rather than immediately after sending it.

Additionally, it would be great to focus on improving the performance of message sending, as it appears to have dropped recently. Enhancing this would also help to reduce that discrepancy.

**Comment by rinej** [comment]
The videos above were after the fix. 
Here is the video from the main branch when we are offline:

https://github.com/user-attachments/assets/f8719613-2c63-48a1-89aa-c8095195db41

As we can see the bug is there as well.

Sorry for confusion :)

**Comment by rinej** [comment]
Yes, there is another bug related to the ticket we have, I was involve in other issues, but now I am going to focus on resolving it

**Comment by rinej** [comment]
I investigated the new bug and here are my findings:
The problem occurs on the main branch.

- The issue occurs on WEB when scrolling through a long list : even after scrolling stops, `ScrollEvents` continue to be triggered (see the attached video).  When this happens, all programmatic calls are blocked. This suggests the issue is related to scrolling, not specifically to message sending.

https://github.com/user-attachments/assets/ba77be6a-9ad8-4163-8fd9-78ac02a816a3

- To eliminate potential causes, I disabled our `scrollToBottom` after sending message logic entirely, but the issue still occurred.

- The list rendering is handled by `InvertedFlatList`, which is built on `FlatList`. Additionally, we have custom scrolling logic implemented within the `InvertedFlatList` component. These are potential areas to investigate further as the sources of bug.

Given that the root cause appears to be different from the original issue described in this thread, and is happening also on `main` branch should we consider opening a separate bug ticket for this? Let me know your thoughts.

**Comment by johncschuster** [comment]
Good thinking, @hoangzinh! I've done that now!

**Comment by johncschuster** [comment]
@rinej let us know when you'll be able to get a PR up for this one!

**Comment by johncschuster** [comment]
@rinej bump!

**Comment by kavimuru** [comment]
Bug is not reproducible in both Hybrid and Standalone app

https://github.com/user-attachments/assets/1306152b-6bcf-489c-ae9f-c6eebb1667a2

**Comment by deetergp** [comment]
Alrighty, I guess we can close this one out then.
